docs/src/user_docs/assembly/execution_contexts.md

```diff
@@ -0,0 +1,105 @@
+
+#### Invoking via `call` and `syscall` instructions
+When a procedure is invoked via a `call` or a `syscall` instruction, the following things happen:
+* Execution moves into a different context. In case of a `call` instruction, a new user context is created. In case of a `syscall` instruction, the execution moves back into the root context.
+* All stack items beyond the 16th item get "hidden" from the invoked procedure. That is, from the standpoint of the invoked procedure the stack depth after a `call` or a `syscall` is 16.
+
+When a procedure returns from a `call` or a `syscall`, the following things happen:
+* Execution moves back to the context from which the procedure was invoked.
+* Stack depth is set to its original depth. Before the stack depth is reset, the VM checks if the current stack depth is exactly 16, and fails otherwise.
+
+Manipulations of the stack depth described above have the following implications:
+- The top 16 elements of the stack can be used to pass parameters and return values between the caller and the callee.
+- Caller's stack beyond the top 16 elements is inaccessible to the callee, and thus, is guaranteed not to change as the result of the call.
+- At the end of its execution, the callee must ensure that stack depth is exactly 16. If this is difficult to ensure manually,  [`truncate_stack`](../stdlib/sys.md) procedure can be used to drop all elements from the stack except for the top 16.
+
+#### Invoking via `exec` instruction
+Procedures invoked via the `exec` instruction, are inlined at their call sites during compilation. Thus, from the standpoint of the final program, executing procedures this way is indistinguishable from manually including procedure code in place of the `exec` instruction. This also means that procedures invoked via the `exec` instruction are executed in the same context as the caller.
+
+### Kernels
+A *kernel* defines a set of procedures which can be invoked from user contexts to be executed in the root context. Miden assembly programs are always compiled against some kernel. The default kernel is empty - i.e., it does not contain any procedures. To compile a program against a non-empty kernel, the kernel needs to be specified when instantiating the [Miden Assembler](https://crates.io/crates/miden-assembly).
+
+A kernel can be defined similarly to a regular [library module](./code_organization.md#library-modules) - i.e., it can have internal and exported procedures. However, there are some small differences between what procedures can do in a kernel module vs. what they can do in a regular library module. Specifically:
+
+- Procedures in a kernel module cannot use `call` or `syscall` instructions. This means that creating a new context from within a syscall is not possible.
+- Unlike procedures in regular library modules, procedures in a kernel module can use `caller` instruction. This instruction puts the hash of the procedure which initiated the parent context onto the stack.
```

Comment:
I am not sure, but I think it must either be `can use 'caller' instructions.` or `can use a  'caller' instruction.`

Reply:
It should be "use the `caller` instruction"